Add math & bitwise ops to U256 type

rostyslavtyshko opened the pull request:
Closes #1798

Braqzen reviewed (commented):
- `u256.sw`
    - Has a commented out import for functions which have been copy and pasted into this module. Could you remove the duplicated functions and re-import them as before?
    - The bitwise traits have the first param as `val: self`. Is there a reason why `self` cannot be used and the `val` part must be specified? It seems to work in the `Shiftable` trait implementation.
    - The `Shiftable` trait has an `else` which contains the `()` type. Afaik this doesn't do anything here. Can it be removed?
    - The `Multiply` trait implementation has a comment about a missing `break` keyword. This is now a feature in sway so you can add that into the `multiply` method (same for `Divide`)
- The tests use the `*` to import. Could you check where that is absolutely needed and change it to explicitly import where it is not?
- Could you also run the formatter using `forc fmt` to format the code?

nfurfaro commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +107,268 @@ impl U256 {
+}
+
+impl core::ops::Shiftable for U256 {
+    fn lsh(self, shift_amount: u64) -> Self {
```
I didn't realize that we would be able to use operators like `<<` if a trait was implemented in `std`. I thought it had to be implemented in `core` for that to work, but I guess because the actual traits are defined in core this works.

nfurfaro reviewed (commented):
A few things I commented on in addition to @Braqzen, but this is looking like it's almost there. good work!

adlerjohn commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +107,268 @@ impl U256 {
     }
+
+    /// Get 4 64 bit words from a single U256 value.
+    fn decompose(self) -> (u64, u64, u64, u64) {
```
Hmm. So ideally this would use from and into, but that requires additional compiler work IIRC. I wouldn't be against just keeping it as-is until then, since it's going away in the not too distant future.

otrho commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +106,263 @@ impl U256 {
+            w2 = shifted_4;
+        } else if w == 3 {
+            w1 = word_4 << b;
+        } else {
+            ();
```
```suggestion
```
Is the final else required by the compiler?

otrho commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +106,263 @@ impl U256 {
+
+        local_res = ~U128::from(0, word_1) + ~U128::from(0, other_word_1) + ~U128::from(0, overflow);
+        let result_a = local_res.lower;
+        if local_res.upper != 0 {
+            // panic on overflow
+            revert(0);
+        };
```
This could probably be an `assert` like below, although we're moving to using `require` instead of `assert` in library code I think.

otrho commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +106,263 @@ impl U256 {
+        let one = ~U256::from(0, 0, 0, 1);
+
+        let mut total = zero;
+        // The algorithm loops <from number of bits - 1> to <zero>.
+        // Need to add 1 here to invalidate the while loop once i == 0 since we
+        // don't have a break keyword.
```
We have `break` and `continue` now.

otrho commented on sway-lib-std/src/vec.sw:
```diff
@@ -14,7 +14,7 @@ struct RawVec<T> {
     /// Create a new `RawVec` with zero capacity.
     fn new() -> Self {
-        Self {
+        RawVec {
```
Are these changes required?  I'm surprised if it used to work but no longer does.  Or is it more about verbosity and self-documentation?

mohammadfawaz replied:
It used to be `RawVec` and was changed to `Self` recently by Emily because she introduced a change that allows the use of `Self` in this context. I suggest we keep it as `Self`.

rostyslavtyshko replied:
Reverted back, as it works now

otrho reviewed (changes requested):
Getting there, some small changes requested and a couple of questions.

I'd also like to see the tests fleshed out some more.  The existing tests could maybe try some bigger numbers, and also we could test for overflow (although I'm not sure how the `should_fail` tests deal with asserts at the moment, they might need to be fleshed out in a different issue/PR).  And tests for `add`, `sub` and shifting should be added.

rostyslavtyshko commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +106,263 @@ impl U256 {
+
+impl core::ops::Ord for U256 {
+    pub fn gt(self, other: Self) -> bool {
+        self.a > other.a || self.a == other.a && self.b > other.b || self.a == other.a && self.b == other.b && self.c > other.c || self.a == other.a && self.b == other.b && self.c == other.c && self.d > other.d
```
It doesn't, and currently formatter has some nasty bugs like changing -= to unworkable - =

rostyslavtyshko commented on sway-lib-std/src/u256.sw:
```diff
@@ -1,7 +1,26 @@
 library u256;
 
 use core::num::*;
+// use core::ops::{lsh_with_carry, rsh_with_carry};
```
After last changes it works fine, so I deleted the copied code and import existing functions

rostyslavtyshko commented:
> Getting there, some small changes requested and a couple of questions.
> 
> I'd also like to see the tests fleshed out some more. The existing tests could maybe try some bigger numbers, and also we could test for overflow (although I'm not sure how the `should_fail` tests deal with asserts at the moment, they might need to be fleshed out in a different issue/PR). And tests for `add`, `sub` and shifting should be added.

Added bigger numbers to mul_test and added a `should_fail` test for overflow. Test cases for `add`, `sub` and shifting  are in the u256_ops_test.

otrho commented on sway-lib-std/src/u256.sw:
```diff
@@ -311,19 +303,19 @@ impl core::ops::Multiply for U256 {
+
+        let mut i = 256 - 1;
+
+        while i >= 0 {
```
`i` is unsigned, so this will always be true.  The `break` below is always ending the loop so this might as well be `while true` to avoid ambiguity.  Personally I'd prefer a var name like `bit_idx` rather than `i` too.
```suggestion
        while true {
```

otrho commented on sway-lib-std/src/u256.sw:
```diff
@@ -341,25 +333,25 @@ impl core::ops::Divide for U256 {
+
+        let mut i = 256 - 1;
+
+        while i >= 0 {
```
Ditto.
```suggestion
        while true {
```

otrho commented on sway-lib-std/src/u256.sw:
```diff
@@ -88,4 +91,286 @@ impl U256 {
+
+impl core::ops::Ord for U256 {
+    pub fn gt(self, other: Self) -> bool {
+        if self.a > other.a {
+            return true
+        } else if self.a == other.a {
+            if self.b > other.b {
+                return true
+            } else if self.b == other.b {
+                if self.c > other.c {
+                    return true
+                } else if self.c == other.c {
+                    if self.d > other.d {
+                        return true
+                    }
+                }
+            }  
+        }
+        false
```
OK, I know I compained about the other long line and suggested nested ifs and return statements, but this isn't great either.  My fault.  But I think this is the same and might be the simplest yet still readable version?
```suggestion
        self.a > other.a
            || (self.a == other.a && self.b > other.b
                || (self.b == other.b && self.c > other.c
                    || (self.c == other.c && self.d > other.d)))
```

otrho reviewed (commented):
Yep, other than simplifying `gt` and `lt` again I think this is OK.

otrho reviewed (approved):
OK, thanks, all good.  Hopefully the CI problem can be fixed, or was maybe an anomaly.

JoshuaBatty commented:
I'm not sure why CI would be failing on this test. It just simulates the json messages when opening a file. 

Perhaps rebasing this PR onto master might help, looks like it hasn't been rebased in 20 days.

JoshuaBatty commented:
hmm still not passing CI after rebasing. It's strange that this test is passing CI for all other PR's except this one.

JoshuaBatty commented:
Looking closer at the diff. Has this actually been rebased onto master properly? Looking at just the changes under the `sway-lsp` folder and there are heaps of changes that are unrelated to this PR at all....

otrho commented:
> Looking closer at the diff. Has this actually been rebased onto master properly?

Considering the age of the original PR (you said it was 20 days old?) I imagine a `git merge` would be better suited than a rebase.  But yeah, it doesn't quite look right.

@rostyslavtyshko It might be worth going back to where you were for the original PR and do a `git merge origin/master` to try and get up to date.

JoshuaBatty commented:
Closing this as the work was already merged in #2445
